src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx

```diff
@@ -0,0 +1,133 @@
+  color: ${props => props.theme.colors.lunesGreyMedium};
+  padding-bottom: 20px;
+`
+
+interface Props {
+  setVisible: (input: boolean) => void
+  setCode: (input: string) => void
+}
+
+const AddCustomDisciplineScreen = ({ setVisible, setCode }: Props): ReactElement => {
+  const appState = useRef(AppState.currentState)
+
+  const [isPressed, setIsPressed] = useState<boolean>(false)
+  const [requestPermissions, setRequestPermissions] = useState<boolean>(false)
+  const [permissionDenied, setPermissionDenied] = useState<boolean>(false)
+
+  const onBarCodeRead = (scanResult: BarCodeReadEvent) => {
+    setCode(scanResult.data)
+    setVisible(false)
+  }
+
+  useEffect(() => {
+    if (requestPermissions) {
+      PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.CAMERA)
+        .then(() => setRequestPermissions(false))
+        .catch(e => console.error(e))
+    }
+  }, [requestPermissions])
+
+  useEffect(() => {
+    const subscription = AppState.addEventListener('change', nextAppState => {
+      if (appState.current.match(/inactive|background/) && nextAppState === 'active') {
+        setRequestPermissions(true)
+      }
+      appState.current = nextAppState
+    })
+    return () => {
+      subscription.remove()
+    }
+  }, [])
+
+  const refresh = () => {
+    Linking.openSettings().catch(() => console.error('Unable to open Settings'))
+  }
+
+  return (
+    <Modal visible transparent animationType='fade'>
+      <Container>
+        <Icon
+          onPress={() => setVisible(false)}
+          onPressIn={() => setIsPressed(true)}
+          onPressOut={() => setIsPressed(false)}
+          activeOpacity={1}>
+          {isPressed ? (
+            <CloseCircleIconBlue testID='close-circle-icon-blue' />
+          ) : (
+            <CloseCircleIconWhite testID='close-circle-icon-white' />
+          )}
+        </Icon>
+        <Camera
+          captureAudio={false}
+          onBarCodeRead={onBarCodeRead}
+          testID='camera'
+          onStatusChange={status => setPermissionDenied(status.cameraStatus === 'NOT_AUTHORIZED')}
+          onCameraReady={() => setPermissionDenied(false)}>
+          {() =>
+            permissionDenied && !requestPermissions ? (
+              <NotAuthorisedView>
+                <NoAuthDescription>
+                  {labels.addCustomDiscipline.qrCodeScanner.noAuthorization.description}
+                </NoAuthDescription>
+                <Button
+                  onPress={() => setVisible(false)}
+                  label={labels.addCustomDiscipline.qrCodeScanner.noAuthorization.back}
+                  buttonTheme={BUTTONS_THEME.outlined}
+                />
+                <Button
+                  onPress={() => refresh()}
+                  label={labels.addCustomDiscipline.qrCodeScanner.noAuthorization.grant}
+                  buttonTheme={BUTTONS_THEME.contained}
+                />
+              </NotAuthorisedView>
```

Comment:
Could we extract this to a separate component?

Reply:
We can, but it think it is not really needed. This component is only < 150 lines long and only has one responsibility: handling the camera and its permissions. So imo it is totally fine to keep it here.

Reply:
I guess it comes down to personal preference, yes. For me this would not be about the general length of the component (I agree here, < 150 LOC is perfectly fine) but the length of the jsx. I think 40 LOC is really long for jsx and makes it hard to review/work with (or at least for me). As this is just a suggestion, feel free to ignore it. Other opinions welcome as well. I just think we should strive for some modularity.

Reply:
oke convinced me, moved it.